### PR TITLE
Provide support for scraping Azure Device Provisioning Service

### LINKED ISSUE #1014: Provide support for scraping Azure Device Provisioning Service
Provide support for scraping Azure Device Provisioning Service

### PR DESCRIPTION
**Pull request by ChristianEder**
Fixes #1014

### CODE REVIEW

**Comment by ChristianEder** [comment]
Tested locally, metrics get provided

File: docs/configuration/v1.x/metrics/device-provisioning-service.md [hunk 1]
```diff
@@ -0,0 +1,34 @@
+---
+layout: default
+title: Azure Device Provisioning Service Declaration
+---
+
+## Azure Device Provisioning Service - ![Availability Badge](https://img.shields.io/badge/Available%20Starting-v1.6-green.svg)
```
Comment (tomkerkhove) [inline docs/configuration/v1.x/metrics/device-provisioning-service.md:6]: Let's align with the official name here:
```suggestion
## Azure IoT Hub Device Provisioning Service (DPS) - ![Availability Badge](https://img.shields.io/badge/Available%20Starting-v1.6-green.svg)
```

File: docs/configuration/v1.x/metrics/device-provisioning-service.md [hunk 2]
```diff
@@ -0,0 +1,34 @@
+---
+layout: default
+title: Azure Device Provisioning Service Declaration
```
Comment (tomkerkhove) [inline docs/configuration/v1.x/metrics/device-provisioning-service.md:3]: Let's align with the official name here:
```suggestion
title: Azure IoT Hub Device Provisioning Service (DPS) Declaration
```

File: changelog/content/experimental/unreleased.md [hunk 3]
```diff
@@ -10,3 +10,4 @@ version: v1.6.0
 - {{% tag added %}} Support for Kubernetes RBAC in Helm chart ([Helm Hub](https://hub.helm.sh/charts/promitor/promitor-agent-scraper) | [#951](https://github.com/tomkerkhove/promitor/issues/951))
 - {{% tag added %}} Support for configuring Pod Security Policy in Helm Chart ([Helm Hub](https://hub.helm.sh/charts/promitor/promitor-agent-scraper) | [#952](https://github.com/tomkerkhove/promitor/issues/952))
 - {{% tag added %}} Support for scraping Azure IoT Hub metrics ([docs](https://promitor.io/configuration/v1.x/metrics/iot-hub) | [#372](https://github.com/tomkerkhove/promitor/issues/372))
+- {{% tag added %}} Support for scraping Azure Device Provisioning Service metrics ([docs](https://promitor.io/configuration/v1.x/metrics/device-provisioning-service) | [#1014](https://github.com/tomkerkhove/promitor/issues/1014))
```
Comment (tomkerkhove) [inline changelog/content/experimental/unreleased.md:13]: Let's align with the official name here:
```suggestion
- {{% tag added %}} Support for scraping Azure IoT Hub Device Provisioning Service (DPS) metrics ([docs](https://promitor.io/configuration/v1.x/metrics/iot-hub-device-provisioning-service) | [#1014](https://github.com/tomkerkhove/promitor/issues/1014))
```

File: docs/configuration/v1.x/metrics/device-provisioning-service.md [hunk 4]
```diff
@@ -0,0 +1,34 @@
+---
```
Comment (tomkerkhove) [inline docs/configuration/v1.x/metrics/device-provisioning-service.md:1]: Let's rename this file to `iot-hub-device-provisioning-service.md`

File: docs/configuration/v1.x/metrics/device-provisioning-service.md [hunk 5]
```diff
@@ -0,0 +1,34 @@
...
+layout: default
+title: Azure Device Provisioning Service Declaration
+---
+
+## Azure Device Provisioning Service - ![Availability Badge](https://img.shields.io/badge/Available%20Starting-v1.6-green.svg)
+
+You can declare to scrape an Azure Device Provisioning Service via the `DeviceProvisioningService` resource type.
```
Comment (tomkerkhove) [inline docs/configuration/v1.x/metrics/device-provisioning-service.md:8]:
```suggestion
You can declare to scrape an Azure IoT Hub Device Provisioning Service via the `DeviceProvisioningService` resource type.
```

File: docs/configuration/v1.x/metrics/device-provisioning-service.md [hunk 6]
```diff
@@ -0,0 +1,34 @@
...
+
+The following fields need to be provided:
+
+- `deviceProvisioningServiceName` - The name of the Azure Device Provisioning Service
```
Comment (tomkerkhove) [inline docs/configuration/v1.x/metrics/device-provisioning-service.md:12]:
```suggestion
- `deviceProvisioningServiceName` - The name of the Azure IoT Hub Device Provisioning Service (DPS)
```

File: docs/configuration/v1.x/metrics/index.md [hunk 7]
```diff
@@ -127,6 +127,7 @@ We also provide a simplified way to scrape the following Azure resources:
 - [Azure Container Registry](container-registry)
 - [Azure Cosmos DB](cosmos-db)
 - [Azure Database for PostgreSQL](postgresql)
+- [Azure Device Provisioning Service](device-provisioning-service)
```
Comment (tomkerkhove) [inline docs/configuration/v1.x/metrics/index.md:130]: Let's align here as well, but order it as well
```suggestion
- [Azure IoT Hub Device Provisioning Service](iot-hub-device-provisioning-service)
```

File: src/Promitor.Agents.Scraper/Validation/MetricDefinitions/ResourceTypes/DeviceProvisioningServiceMetricValidator.cs [hunk 8]
```diff
@@ -0,0 +1,25 @@
...
+    {
+        public IEnumerable<string> Validate(MetricDefinition metricDefinition)
+        {
+            Guard.NotNull(metricDefinition, nameof(metricDefinition));
+
+            foreach (var resourceDefinition in metricDefinition.Resources.Cast<DeviceProvisioningServiceResourceDefinition>())
+            {
+                if (string.IsNullOrWhiteSpace(resourceDefinition.DeviceProvisioningServiceName))
+                {
+                    yield return "No Azure Device Provisioning Service name is configured";
```
Comment (tomkerkhove) [inline src/Promitor.Agents.Scraper/Validation/MetricDefinitions/ResourceTypes/DeviceProvisioningServiceMetricValidator.cs:20]:
```suggestion
                    yield return "No Azure IoT Hub Device Provisioning Service name is configured";
```

**Comment by ChristianEder** [comment]
Resolved your rename remarks. Yes I have an example in action, but its running locally on my machine (the promitor container) and inside my Azure Subscription (the IoT Hub & DPS) - how do you want me to share?

**Comment by tomkerkhove** [comment]
> Resolved your rename remarks. Yes I have an example in action, but its running locally on my machine (the promitor container) and inside my Azure Subscription (the IoT Hub & DPS) - how do you want me to share?

Great thanks!

Maybe the output of the scraping endpoint if you can share it?

**Comment by ChristianEder** [comment]
\#  HELP azure_dps_attestations Amount of attestations
\# TYPE azure_dps_attestations gauge
azure_dps_attestations{resource_group="ched",subscription_id="c24aae66-cc75-4de5-8148-c83a5981c88f",resource_uri="subscriptions/c24aae66-cc75-4de5-8148-c83a5981c88f/resourceGroups/ched/providers/Microsoft.Devices/provisioningServices/mypromitordps",instance_name="mypromitordps"} 1
\# HELP promitor_ratelimit_arm Indication how many calls are still available before Azure Resource Manager is going to throttle us.
\# TYPE promitor_ratelimit_arm gauge
promitor_ratelimit_arm{tenant_id="1a7dacf1-7fd4-45b1-bf70-01265d702d04",subscription_id="c24aae66-cc75-4de5-8148-c83a5981c88f",app_id="9d78938a-034b-429f-ae35-c57b969e080f"} 11990`

**Pull request merged**